### ISSUE TITLE
Resource abstraction (React renderer), how to handle updates?

### ISSUE THREAD

**Issue by LevelbossMike** [starbeamjs/starbeam#76]
I'm trying to build a resource that encapsulates an XState interpreter- much like the ember-statecharts add-on provides. Setting up the resource works as expected, but I can't figure out how to set up a reactive update pattern. How would I do what ember-statecharts is doing in starbeam? https://github.com/LevelbossMike/ember-statecharts/blob/master/ember-statecharts/src/resources/statechart.ts#L172

__What I want to do:__
The first time the resource gets setup, I set up a stateful thing (in this case, an XState interpreter). Then when things that I reference when I use the resource change, I would like to provide an API to react to that state change, i.e. send an event or restart the underlying interpreter.

```js
// this is what I have as a resource
import { Resource, Cell, Formula } from '@starbeam/universal';
import { interpret } from 'xstate';

export const Statechart = function({ machine }) {
  return Resource((r) => {
    const state = Cell(machine.initialState);

    const interpreter = interpret(machine)

    interpreter.start().onTransition((newState) => {
      state.set(newState);
    })


    r.on.cleanup(() => {
      interpreter.stop();
    })

    return Formula(() => {
      return { 
        state: state.current,
        send: interpreter.send
      }
    })
  })
}
```

I use it like this in my react test app. This code example is just to showcase that setting up the resource works as expected:

```jsx
import React from 'react';
import { use } from '@starbeam/react';
import { Statechart } from '../resources/statechart';
import { createMachine } from 'xstate';

const toggleMachine = createMachine({
  initial: 'off',
  states: {
    on: {
      on: {
        TOGGLE: 'off'
      }
    },
    off: {
      on: {
        TOGGLE: 'on'
      }
    }
  }
});

function Toggle() {
  const statechart = use(() => Statechart({ machine: toggleMachine }), [])
  
  return (
    <div>
      <div>
        {statechart?.state.value}
      </div>
      <button type="button" onClick={() => statechart.send('TOGGLE')}>TOGGLE</button>
    </div>
  )
}

export default Toggle;
```

I wonder how I could make the following work: I want to provide an update handler for the Statechart resource instead of tearing it down and restarting it. This means I need to keep the interpreter around and let the user decide how they would like to handle the update:

```jsx
<button type="button" onClick={() => /* update machine property that gets consumed in use */>UPDATE</button>
```

I thought this might work, but it doesn't call my resource handler with the updated value from `newMachine`:

```jsx
function Toggle() {
  const machine = useSetup(() => Cell(toggleMachine))
  
  return useReactive(() => {
    const statechart = use(() => Statechart({ machine: machine.current }), [machine.current])
    return (
      <div>
        <div>
          {statechart?.state.value}
        </div>
        <button type="button" onClick={() => statechart.send('TOGGLE')}>TOGGLE</button>
        <button type="button" onClick={() => machine.update(() => {
          const newMachine = toggleMachine.withContext({ count: 1})
          return newMachine;
          })}>
          Update
        </button>
        <div>
          {machine.current.context.count}
        </div>
      </div>
    )

  }
  ) 
}
```

Here's a Stackblitz reproduction of what I'm trying to do:

https://stackblitz.com/edit/starbeam-react-base-rp6irc?file=src/resources/statechart.js

When the application boots up, you should see a log where the statechart resources sets up. The `context.count` should be 0. you can then click the update button, which updates the machine cell referenced in the use function. You will then see a log that makes sure the `newMachine` actually has a `context.count` of 1. I would then imagine the resource being set up with that machine, the function runs again but with the old machine, apparently without an updated `context.count` property.

Based on our discussions in Discord this might be a bug. Thanks for looking into this @wycats 🙇 👏

**Comment by wycats** [comment]
This was fixed in e25b96779711f2257b0aec983d6b534b38ed969d and released in 0.8.7.

**Comment by wycats** [comment]
Thanks for the report and sticking with it @LevelbossMike!
